### PR TITLE
Add `--lockfile_mode=refresh`

### PR DESCRIPTION
**Pull request by fmeum**
RELNOTES: The new `refresh` value for `--lockfile_mode` behaves like the `update` mode, but additionally forces a refresh of mutable registry content (yanked versions and missing module versions) when switched to or from time to time while enabled.

### CODE REVIEW

**Comment by fmeum** [comment]
I added the refresh logic. I didn't find a way to test it that would require making the timeout configurable, but I verified locally that it works with shorter durations.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/IndexRegistry.java [hunk 1]
```diff
@@ -60,6 +60,7 @@ public class IndexRegistry implements Registry {
   public enum KnownFileHashesMode {
     IGNORE,
     USE_AND_UPDATE,
+    USE_IMMUTABLE_AND_UPDATE,
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/IndexRegistry.java:63]: I think these have gotten complicated enough to warrant documentation here.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/RegistryFunction.java [hunk 2]
```diff
@@ -17,16 +17,22 @@
...
 import com.google.devtools.build.skyframe.SkyValue;
 import java.net.URISyntaxException;
+import java.time.Duration;
 import javax.annotation.Nullable;
 
 /** A simple SkyFunction that creates a {@link Registry} with a given URL. */
 public class RegistryFunction implements SkyFunction {
+  public static final Precomputed<Long> LAST_INVALIDATION =
+      new Precomputed<>("last_registry_invalidation");
+  public static final Duration INVALIDATION_INTERVAL = Duration.ofHours(1);
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/RegistryFunction.java:34]: please add javadoc

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/RegistryFunction.java [hunk 3]
```diff
@@ -41,6 +47,12 @@ public SkyValue compute(SkyKey skyKey, Environment env)
       throws InterruptedException, RegistryException {
     LockfileMode lockfileMode = BazelLockFileFunction.LOCKFILE_MODE.get(env);
 
+    if (lockfileMode == LockfileMode.REFRESH) {
+      // Set to the current time in BazelRepositoryFunction after INVALIDATION_INTERVAL has passed.
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/RegistryFunction.java:51]: s/BazelRepositoryFunction/BazelRepositoryModule

File: src/main/java/com/google/devtools/build/lib/bazel/BazelRepositoryModule.java [hunk 4]
```diff
@@ -526,6 +530,22 @@ public void beforeCommand(CommandEnvironment env) throws AbruptExitException {
       starlarkRepositoryFunction.setRepositoryRemoteExecutor(remoteExecutor);
       singleExtensionEvalFunction.setRepositoryRemoteExecutor(remoteExecutor);
       delegatingDownloader.setDelegate(env.getRuntime().getDownloaderSupplier().get());
+
+      clock = env.getClock();
+      try {
+        var lastYankedVersionsInvalidationValue =
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/BazelRepositoryModule.java:536]: lastRegistryInvalidationValue?

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/RegistryFunction.java [hunk 5]
```diff
@@ -17,16 +17,22 @@
...
 import com.google.devtools.build.skyframe.SkyFunctionException;
 import com.google.devtools.build.skyframe.SkyKey;
 import com.google.devtools.build.skyframe.SkyValue;
 import java.net.URISyntaxException;
+import java.time.Duration;
 import javax.annotation.Nullable;
 
 /** A simple SkyFunction that creates a {@link Registry} with a given URL. */
 public class RegistryFunction implements SkyFunction {
+  public static final Precomputed<Long> LAST_INVALIDATION =
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/RegistryFunction.java:32]: can this just be a `Precomputed<Instant>`?

**Review by Wyverald — APPROVED** [review]
LGTM. just need some documentation updates etc

**Comment by fmeum** [comment]
@bazel-io fork 7.2.0

**Review by Wyverald — APPROVED** [review]

**Pull request closed without merging**